drivers/ds75lx/ds75lx.c

```diff
@@ -0,0 +1,148 @@
+#include "debug.h"
+
+#define DEV_I2C      (dev->params.i2c)
+#define DEV_ADDR     (dev->params.addr)
+
+/* Mapping table between temperature resolution and max conversion time. */
+static const uint8_t _max_conversion_times[4] = {
+    [DS75LX_RESOLUTION_9]  = DS75LX_MAX_CONVERSION_25,
+    [DS75LX_RESOLUTION_10] = DS75LX_MAX_CONVERSION_50,
+    [DS75LX_RESOLUTION_11] = DS75LX_MAX_CONVERSION_100,
+    [DS75LX_RESOLUTION_12] = DS75LX_MAX_CONVERSION_250,
+};
+
+static int _update_configuration_bits(const ds75lx_t *dev, uint8_t bit,
+                                      uint8_t mask, bool set)
+{
+    /* Acquire exclusive access */
+    i2c_acquire(DEV_I2C);
+
+    uint8_t config;
+    if (i2c_read_reg(DEV_I2C, DEV_ADDR, DS75LX_REG_CONFIGURATION, &config, 0) < 0) {
+        DEBUG("[ds75lx] error reading configuration register\n");
+        /* Release I2C device */
+        i2c_release(DEV_I2C);
+
+        return -DS75LX_ERR_I2C;
+    }
+
+    DEBUG("[ds75lx] initial configuration register value: 0x%02X\n", config);
+
+    /* clear bits */
+    config &= ~mask;
+
+    /* set bits if required */
+    if (set) {
+        config |= bit;
+    }
+
+    DEBUG("[ds75lx] configuration register value: 0x%02X\n", config);
+
+    if (i2c_write_reg(DEV_I2C, DEV_ADDR, DS75LX_REG_CONFIGURATION, config, 0) < 0) {
+        DEBUG("[ds75lx] error writing configuration register\n");
+        /* Release I2C device */
+        i2c_release(DEV_I2C);
+
+        return -DS75LX_ERR_I2C;
+    }
+
+    /* Release I2C device */
+    i2c_release(DEV_I2C);
+
+    return DS75LX_OK;
+}
+
+int ds75lx_init(ds75lx_t *dev, const ds75lx_params_t *params)
+{
+    dev->params = *params;
+
+    /* Set resolution bits + force shutdown of sensor */
+    return _update_configuration_bits(dev,
+        (dev->params.resolution << DS75LX_CONF_R0_POS) | (1 << DS75LX_CONF_SD_POS),
+        (DS75LX_CONF_R0_MASK | (1 << DS75LX_CONF_SD_POS)), true);
+}
+
+int ds75lx_read_temperature(const ds75lx_t *dev, int16_t *temperature)
+{
+    /* Acquire exclusive access */
+    i2c_acquire(DEV_I2C);
+
+    uint8_t tmp[2];
+    uint16_t temp;
+
+    if (i2c_read_regs(DEV_I2C, DEV_ADDR, DS75LX_REG_TEMPERATURE, tmp, 2, 0) < 0) {
+        DEBUG("[ds75lx] error reading temperature register\n");
+        /* Release I2C device */
+        i2c_release(DEV_I2C);
+
+        return -DS75LX_ERR_I2C;
+    }
+
+    /* Release I2C device */
+    i2c_release(DEV_I2C);
+
+    temp = (tmp[0] << 8) | tmp[1];
+    DEBUG("[ds75lx] temperature register content 0x%04X\n", temp);
+
+    /* isolate integer part of the temperature */
+    int8_t temp_int = (temp & 0xff00) >> 8;
+    /* compute fractional part of the temperature, the LSB bits 3 to 0 are
+    always zero and not used in the conversion */
+    uint8_t temp_frac = (temp & 0x00f0) >> 4;
+
+    /* fractional part is a multiple of 0.0625. Temperature is returned in c°C */
+    *temperature  = (temp_int * 100 + ((uint16_t)temp_frac * 100 >> 4));
```

Comment:
Not sure if I'm missing something but if temp_frac is b'1000 this converts to 50 or 0.5, but if its b'1111 this converts to 93.75 or 0.9375 which is incorrect since it should be 0.0625.  Shouldn't you be using 2 complement of the fraction part  (since you have been using unsigned types for the fractional part)

` uint8_t temp_frac = ((~temp & 0x00f0) >> 4) + 1;`

Reply:
No because only integral part is 2 complemented. Fractional part is always positive.

For example, for `-25.0625 °C` DS75LX returns `0xE6F0`.  Here integral part is `0xE6` (which is -26 in 2-complement representation) and fractional part is `0b1111` (which represents 0.9375).
We can verify that **-26 + 0.9375 = -25.0625**.